packages/firestore/src/util/async_queue.ts

```diff
@@ -208,12 +218,65 @@ export class AsyncQueue {
     this.enqueue(op);
   }
 
+  /**
+   * Regardless if the queue has initialized shutdown, adds a new operation to the
+   * queue without waiting for it to complete (i.e. we ignore the Promise result).
+   */
+  enqueueAndForgetEvenAfterShutdown<T extends unknown>(
+    op: () => Promise<T>
+  ): void {
+    this.verifyNotFailed();
+    // tslint:disable-next-line:no-floating-promises
+    this.enqueueInternal(op);
+  }
+
+  /**
+   * Regardless if the queue has initialized shutdown, adds a new operation to the
+   * queue.
+   */
+  private enqueueEvenAfterShutdown<T extends unknown>(
+    op: () => Promise<T>
+  ): Promise<T> {
+    this.verifyNotFailed();
+    // tslint:disable-next-line:no-floating-promises
+    return this.enqueueInternal(op);
+  }
+
+  /**
+   * Adds a new operation to the queue and initialize the shut down of this queue.
+   * Returns a promise that will be resolved when the promise returned by the new
+   * operation is (with its value).
+   * Once this method is called, the only possible way to request running an operation
+   * is through `enqueueAndForgetEvenAfterShutdown`.
+   */
+  enqueueAndInitilizeShutdown<T extends unknown>(
+    op: () => Promise<T>
+  ): Promise<T> {
+    this.verifyNotFailed();
+    if (this._isShuttingDown) {
+      // Return a Promise resolves right away if it is already shutdown.
+      return new Promise<T>(resolve => resolve(undefined));
```

Comment:
You would normally use `Promise.resolve()` for this...  but I see that gives a type error because `undefined` isn't convertible to `T`, which is a valid complaint. So I think if this is the behavior we want, the right thing to do here is `return Promise.resolve(undefined as T)` but we are subverting the type system (T likely isn't supposed to include `undefined`).

But as it happens, the only code we pass to `enqueueAndInitilizeShutdown()` returns `Promise<void>` so we should probably just make this method non-generic, which solves our problems.  Finally, you could:
1. Make this method `async` so you can use `await` and the compiler will automatically make all code paths return `Promise` (i.e. it'll generate a `Promise.resolve(undefined)` for the no-op case.
2. Use your `enqueueEvenAfterShutdown()` so you don't need to do the enqueueInternal() / isShuttingDown=true reverse ordering trick.

So you end up with:
```
  async enqueueAndInitilizeShutdown(op: () => Promise<void>): Promise<void> {
    this.verifyNotFailed();
    if (!this._isShuttingDown) {
      this._isShuttingDown = true;
      await this.enqueueEvenAfterShutdown(op);
    }
  }
```

This will cause an error in your async_queue.test.ts test which you can fix by changing:
```
    queue.enqueueAndInitilizeShutdown(() => doStep(2));
```
to:
```
    queue.enqueueAndInitilizeShutdown(async () => { doStep(2); });
```

or you could just change doStop() to return `Promise<void>` (as I mention again below).